core/src/main/java/com/yandex/yoctodb/util/mutable/impl/FoldedByteArrayIndexedList.java

```diff
@@ -0,0 +1,179 @@
+/**
+ * {@link ByteArrayIndexedList} with variable sized elements
+ *
+ * @author irenkamalova
+ */
+@NotThreadSafe
+public final class FoldedByteArrayIndexedList
+        implements ByteArrayIndexedList {
+    @NotNull
+    private final Set<UnsignedByteArray> elements;
+    @NotNull
+    private final List<Long> offsets;
+    @NotNull
+    private final SortedMap<Integer, Integer> docIdOffsetIndex;
+    private final int sizeOfIndexOffsetValue; // how many bites
+    private final int databaseDocumentsCount;
+
+
+    public FoldedByteArrayIndexedList(
+            @NotNull final Map<UnsignedByteArray, LinkedList<Integer>> elements,
+            int databaseDocumentsCount) {
+
+        this.databaseDocumentsCount = databaseDocumentsCount;
+        this.docIdOffsetIndex = new TreeMap<>();
+        this.offsets = new ArrayList<>();
+        this.elements = elements.keySet();
+
+        long elementOffset = 0;
+        // reserve first value for empty documents
+        offsets.add(-1L);
+
+        for (Map.Entry<UnsignedByteArray, LinkedList<Integer>> elem :
+                elements.entrySet()) {
+            UnsignedByteArray value = elem.getKey();
+            offsets.add(elementOffset);
+
+            final long currentElementOffset = elementOffset;
+            elem.getValue().forEach( docId ->
+                    docIdOffsetIndex
+                         .put(docId,
+                         offsets.indexOf(currentElementOffset)));
+            elementOffset += value.getSizeInBytes();
+        }
+
+        offsets.add(elementOffset);
+        offsets.set(0, 0L);
+
+        // analyze result of offsets
+        final int offsetCount = offsets.size();
+        if (offsetCount < (1 << Byte.SIZE)) {
+            sizeOfIndexOffsetValue = Byte.BYTES;
+        } else if (offsetCount < (1 << Short.SIZE)) {
+            sizeOfIndexOffsetValue = Short.BYTES;
+        } else {
+            sizeOfIndexOffsetValue = Integer.BYTES;
+        }
+
+    }
+
+    @Override
+    public long getSizeInBytes() {
+        long elementSize = 0;
+        for (UnsignedByteArray element : elements) {
+            elementSize += element.length();
+        }
+
+        return Integer.BYTES + // Element count in bytes
+                Integer.BYTES + // offsets count in bytes
+                sizeOfIndexOffsetValue *
+                        databaseDocumentsCount + // indexes offsets in bytes
+                Long.BYTES * offsets.size() + // offsets array size in bytes
+                elementSize; // Element array size in bytes
+    }
+
+    @Override
+    public void writeTo(
+            @NotNull final OutputStream os) throws IOException {
+        // elements count
+        os.write(Ints.toByteArray(databaseDocumentsCount));
+
+        // write offsets count!
+        // before writing indexes because
+        // you should to know how to deserialize it
+        os.write(Ints.toByteArray(offsets.size()));
+
+        List<Integer> offsetIndexes = new ArrayList<>();
```

Comment:
`int[]` would definitely be better here. And once again please think about changing `docIdOffsetIndex` to `int[]`.